// pg-protocol/dist/outbound-serializer.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const assert_1 = __importDefault(require("assert"));
const serializer_1 = require("./serializer");
const buffer_list_1 = __importDefault(require("./testing/buffer-list"));
describe('serializer', () => {
    it('builds startup message', function () {
        const actual = serializer_1.serialize.startup({
            user: 'brian',
            database: 'bang',
        });
        assert_1.default.deepEqual(actual, new buffer_list_1.default()
            .addInt16(3)
            .addInt16(0)
            .addCString('user')
            .addCString('brian')
            .addCString('database')
            .addCString('bang')
            .addCString('client_encoding')
            .addCString('UTF8')
            .addCString('')
            .join(true));
    });
    it('builds password message', function () {
        const actual = serializer_1.serialize.password('!');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('!').join(true, 'p'));
    });
    it('builds request ssl message', function () {
        const actual = serializer_1.serialize.requestSsl();
        const expected = new buffer_list_1.default().addInt32(80877103).join(true);
        assert_1.default.deepEqual(actual, expected);
    });
    it('builds SASLInitialResponseMessage message', function () {
        const actual = serializer_1.serialize.sendSASLInitialResponseMessage('mech', 'data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString('mech').addInt32(4).addString('data').join(true, 'p'));
    });
    it('builds SCRAMClientFinalMessage message', function () {
        const actual = serializer_1.serialize.sendSCRAMClientFinalMessage('data');
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addString('data').join(true, 'p'));
    });
    it('builds query message', function () {
        const txt = 'select * from boom';
        const actual = serializer_1.serialize.query(txt);
        assert_1.default.deepEqual(actual, new buffer_list_1.default().addCString(txt).join(true, 'Q'));
    });
    describe('parse message', () => {
        it('builds parse message', function () {
            const actual = serializer_1.serialize.parse({ text: '!' });
            const expected = new buffer_list_1.default().addCString('').addCString('!').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
        it('builds parse message with named query', function () {
            const actual = serializer_1.serialize.parse({
                name: 'boom',
                text: 'select * from boom',
                types: [],
            });
            const expected = new buffer_list_1.default().addCString('boom').addCString('select * from boom').addInt16(0).join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
        it('with multiple parameters', function () {
            const actual = serializer_1.serialize.parse({
                name: 'force',
                text: 'select * from bang where name = $1',
                types: [1, 2, 3, 4],
            });
            const expected = new buffer_list_1.default()
                .addCString('force')
                .addCString('select * from bang where name = $1')
                .addInt16(4)
                .addInt32(1)
                .addInt32(2)
                .addInt32(3)
                .addInt32(4)
                .join(true, 'P');
            assert_1.default.deepEqual(actual, expected);
        });
    });
    describe('bind messages', function () {
        it('with no values', function () {
            const actual = serializer_1.serialize.bind();
            const expectedBuffer = new buffer_list_1.default()
                .addCString('')
                .addCString('')
                .addInt16(0)
                .addInt16(0)
                .addInt16(1)
                .addInt16(0)
                .join(true, 'B');
            assert_1.default.deepEqual(actual, expectedBuffer);
        });
        it('with named statement, portal, and
